x-pack/metricbeat/module/aws/ec2/ec2.go

```diff
@@ -0,0 +1,330 @@
+	}
+
+	for _, regionName := range regionsList {
+		cfg.Region = regionName
+		svcEC2 := ec2.New(cfg)
+		instanceIDs, instancesOutputs, err := getInstancesPerRegion(svcEC2)
+		if err != nil {
+			report.Error(errors.Wrap(err, "getInstancesPerRegion failed"))
+		}
+
+		svcCloudwatch := cloudwatch.New(cfg)
+		for _, instanceID := range instanceIDs {
+			init := true
+			getMetricDataOutput := &cloudwatch.GetMetricDataOutput{NextToken: nil}
+			for init || getMetricDataOutput.NextToken != nil {
+				init = false
+				output, err := getMetricDataPerRegion(instanceID, getMetricDataOutput.NextToken, svcCloudwatch)
+				if err != nil {
+					report.Error(errors.Wrap(err, "getMetricDataPerRegion failed"))
+				}
+				getMetricDataOutput.MetricDataResults = append(getMetricDataOutput.MetricDataResults, output.MetricDataResults...)
+			}
+			reportCloudWatchEvents(getMetricDataOutput, instanceID, instancesOutputs[instanceID], regionName, report)
+		}
+	}
+}
+
+// MockFetch methods implements the data gathering and data conversion using MockEC2Client and MockCloudWatchClient
+func (m *MetricSet) MockFetch(report mb.ReporterV2) {
+	svcEC2Mock := &MockEC2Client{}
+	instanceIDs, instancesOutputs, err := getInstancesPerRegion(svcEC2Mock)
+	if err != nil {
+		report.Error(errors.Wrap(err, "getInstancesPerRegion failed"))
+	}
+
+	svcCloudwatchMock := &MockCloudWatchClient{}
+	for _, instanceID := range instanceIDs {
+		getMetricDataOutput, err := getMetricDataPerRegion(instanceID, nil, svcCloudwatchMock)
+		if err != nil {
+			report.Error(errors.Wrap(err, "getMetricDataPerRegion failed"))
+		}
+		reportCloudWatchEvents(getMetricDataOutput, instanceID, instancesOutputs[instanceID], "us-west-1", report)
+	}
+}
+
+func getRegions(svc ec2iface.EC2API) (regionsList []string, err error) {
+	input := &ec2.DescribeRegionsInput{}
+	req := svc.DescribeRegionsRequest(input)
+	output, err := req.Send()
+	if err != nil {
+		fmt.Println("Failed DescribeRegions: ", err)
+		return
+	}
+	for _, region := range output.Regions {
+		regionsList = append(regionsList, *region.RegionName)
+	}
+	return
+}
+
+func reportCloudWatchEvents(getMetricDataOutput *cloudwatch.GetMetricDataOutput, instanceID string, instanceOutput ec2.Instance, regionName string, report mb.ReporterV2) {
+	machineType, err := instanceOutput.InstanceType.MarshalValue()
+	if err != nil {
+		report.Error(errors.Wrap(err, "instance.InstanceType.MarshalValue failed"))
+	}
+
+	event := mb.Event{}
+	event.RootFields = common.MapStr{}
+	event.RootFields.Put("service.name", aws.ModuleName)
+	event.ModuleFields = common.MapStr{}
+	event.ModuleFields.Put("cloud.provider", "ec2")
+	event.ModuleFields.Put("cloud.instance.id", instanceID)
+	event.ModuleFields.Put("cloud.machine.type", machineType)
+	event.ModuleFields.Put("cloud.availability_zone", *instanceOutput.Placement.AvailabilityZone)
+	event.ModuleFields.Put("cloud.image.id", *instanceOutput.ImageId)
+	event.ModuleFields.Put("cloud.region", regionName)
```

Comment:
All these `cloud.*` should be Put under `event.RootFields`, so they end up matching with what `add_cloud_metadata` provides